### PR TITLE
Integrate @adguard/tsurlfilter

### PR DESCRIPTION
**Pull request by mjethani**
Let's integrate the [@adguard/tsurlfilter](https://www.npmjs.com/package/@adguard/tsurlfilter) library into the benchmarks.

### CODE REVIEW

**Comment by mjethani** [comment]
@ameshkov

To try this out:

```
git clone -b integrate-tsurlfilter https://github.com/mjethani/adblocker
cd adblocker/packages/adblocker-benchmarks
npm install
make tsurlfilter
```

File: packages/adblocker-benchmarks/blockers/tsurlfilter.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+const {
+  Engine,
+  Request,
+  RequestType,
+  RuleStorage,
+  StringRuleList,
+  setConfiguration,
+} = require('@adguard/tsurlfilter/dist/bundle.min.cjs');
```
Comment (mjethani) [inline packages/adblocker-benchmarks/blockers/tsurlfilter.js:16]: @ameshkov we noticed that minifying code made a huge difference to the initialization time. We are minifying all the main content blockers before running the benchmarks.

File: packages/adblocker-benchmarks/Makefile [hunk 2]
```diff
@@ -56,6 +64,9 @@ ublock:
 adblockplus:
 	$(node_command) run.js adblockplus $(run_options)
 
+tsurlfilter:
```
Comment (remusao) [inline packages/adblocker-benchmarks/Makefile:67]: Nit. Can we maybe call this `adguard` to make it clearer?

File: packages/adblocker-benchmarks/blockers/tsurlfilter.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  [ 'stylesheet', RequestType.Stylesheet ],
+  [ 'image', RequestType.Image ],
+  [ 'xmlhttprequest', RequestType.XmlHttpRequest ],
+  [ 'media', RequestType.Media ],
+  [ 'font', RequestType.Font ],
+  [ 'other', RequestType.Other ],
+]);
+
+module.exports = class TSUrlFilter {
+  static async initialize() {
```
Comment (mjethani) [inline packages/adblocker-benchmarks/blockers/tsurlfilter.js:30]: @@ameshkov the `initialize()` function is supposed to measure any global initialization for the library. I thought it was appropriate to call `setConfiguration()` here.

File: packages/adblocker-benchmarks/blockers/tsurlfilter.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    setConfiguration({
+      engine: 'extension',
+      version: '1.0.0',
+      verbose: true,
+    });
+  }
+
+  static async parse(rawLists) {
+    const list = new StringRuleList(1, rawLists, true, true, true);
```
Comment (mjethani) [inline packages/adblocker-benchmarks/blockers/tsurlfilter.js:39]: @ameshkov I'm passing `true` for all three options. I'm not sure if it makes a difference. In any case, all the filters we are passing here are URL filters (a subset of EasyList). Let me know if you see anything wrong with this.

File: packages/adblocker-benchmarks/blockers/tsurlfilter.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  constructor(engine) {
+    this.engine = engine;
+  }
+
+  match({ url, frameUrl, type }) {
+    const requestType = requestTypes.get(type) || RequestType.Other;
+    const request = new Request(url, frameUrl, requestType);
+    const result = this.engine.matchRequest(request);
```
Comment (mjethani) [inline packages/adblocker-benchmarks/blockers/tsurlfilter.js:51]: @ameshkov I notice that the filter matching is quite slow. In the API, there's a function for returning _all_ matches and then another function for returning only the basic match. Ideally there should be one function for returning the basic match, and the search should be terminated when a match is found. I hope I'm not using the wrong function here. Let me know if there's already a better way to do this.

File: packages/adblocker-benchmarks/blockers/tsurlfilter.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    const basicResult = result.getBasicResult();
+    return basicResult && !basicResult.whitelist;
```
Comment (mjethani) [inline packages/adblocker-benchmarks/blockers/tsurlfilter.js:53]: @ameshkov I'm checking that the `whitelist` property is `false`. Let me know if this is wrong. I do see that the library blocks approximately the same number of requests as the other content blockers.

**Comment by ameshkov** [comment]
```
Submodule path 'blockers/ublock/submodules/uAssets': checked out '9df535fe7955996069377cbc63631a51aaeb72cd'
make -C ./blockers/ublock install-nodejs clean
tools/make-nodejs.sh
tools/make-nodejs.sh: line 61: eslint: command not found
make[1]: *** [dist/build/uBlock0.nodejs] Error 127
make: *** [node_modules/@gorhill/ubo-core] Error 2
error Command failed with exit code 2.
```

Requires globally installed eslint?

File: packages/adblocker-benchmarks/Makefile (hunk 2 above)
Comment (mjethani) [reply packages/adblocker-benchmarks/Makefile:67]: >Can we maybe call this `adguard` to make it clearer?

It's not used in the AdGuard browser extension yet. I thought we should just call it by the name of the library for now. When it's integrated into the AdGuard browser extension, we can change the name here.

**Comment by mjethani** [comment]
>Requires globally installed eslint?

Sorry about this. The @gorhill/ubo-core integration is also a work in progress. Yes, it requires global ESLint.

File: packages/adblocker-benchmarks/blockers/tsurlfilter.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjethani) [reply packages/adblocker-benchmarks/blockers/tsurlfilter.js:39]: I'm not sure exactly what these arguments are either, and there's not much documentation. Let's wait for @ameshkov to clarify (https://github.com/cliqz-oss/adblocker/pull/2109#discussion_r682790331), and then I'll add some comments.

Comment (ameshkov) [reply packages/adblocker-benchmarks/blockers/tsurlfilter.js:39]: No worries, it's still a work in progress at this point, we're only entering the performance tuning stage.

File: packages/adblocker-benchmarks/blockers/tsurlfilter.js (hunk 6 above)
Comment (ameshkov) [reply packages/adblocker-benchmarks/blockers/tsurlfilter.js:53]: Yep, that's right.

File: packages/adblocker-benchmarks/blockers/tsurlfilter.js (hunk 5 above)
Comment (ameshkov) [reply packages/adblocker-benchmarks/blockers/tsurlfilter.js:51]: @mjethani No worries, it's still a work in progress at this point, we're only entering the performance tuning stage.

**Comment by mjethani** [comment]
@ameshkov use `make tsurlfilter SHOW_MEMORY=1` to see information about memory usage (work in progress!). Use `make brave`, `make ublock`, `make adblockplus`, and `make cliqz` to run benchmarks for the respective content blocker.

File: packages/adblocker-benchmarks/blockers/tsurlfilter.js (hunk 4 above)
Comment (ameshkov) [reply packages/adblocker-benchmarks/blockers/tsurlfilter.js:39]: Yes, that's okay and it indeed does not make any difference.

**Comment by ameshkov** [comment]
> @ameshkov use make tsurlfilter SHOW_MEMORY=1 to see information about memory usage (work in progress!). Use make brave, make ublock, make adblockplus, and make cliqz to run benchmarks for the respective content blocker.

Great, thank you, that's a great tool and we'll definitely be using it to measure the impact of the new changes.

On a side note, it probably makes sense to add info about requirements somewhere, I noticed it also requires rust.

File: packages/adblocker-benchmarks/blockers/tsurlfilter.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjethani) [reply packages/adblocker-benchmarks/blockers/tsurlfilter.js]: @remusao I've removed the boolean arguments and documented the numerical one.

File: packages/adblocker-benchmarks/Makefile (hunk 2 above)
Comment (mjethani) [reply packages/adblocker-benchmarks/Makefile:67]: >It's not used in the AdGuard browser extension yet. I thought we should just call it by the name of the library for now. When it's integrated into the AdGuard browser extension, we can change the name here.

@remusao I'll make this as resolved but let me know if you disagree.

**Comment by mjethani** [comment]
> … we'll definitely be using it to measure the impact of the new changes.

I can second this. I used it a lot for the optimization of Adblock Plus.

> … probably makes sense to add info about requirements

Thanks! I'm trying to address the ESLint issue: https://github.com/gorhill/uBlock/pull/3803 As for Rust, I guess we'll have to mention it in the requirements.

**Comment by mjethani** [comment]
@remusao I guess it looks good to @ameshkov so I've marked it as ready for review. Let me know if you have any comments and then we can merge it.

**Review by remusao — APPROVED** [review]

**Comment by mjethani** [comment]
@ameshkov @remusao thanks for your time!

Note that we are also benchmarking serialization and deserialization and I have skipped this for @adguard/tsurlfilter to keep it simple for now.

**Pull request merged**